### PR TITLE
Handle deferred values uniformly via `DeferredUtil`

### PR DESCRIPTION
**Pull request by bamboo**
In preparation to also accept Kotlin lambdas as deferred computations, this PR makes their handling more uniform across the board by:
- Treating deferred computation represented by `Callable` instances the same as `Closure` instances (`Closure` is `Callable`), this makes the code simpler and the error handling more uniform;
- Reuse `DeferredUtil.unpack` for code that handles `Callable`s as deferred computations, this extends their reach to `Factory` and `Provider` and later to Kotlin's `Function0` (lambda expressions)

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/runtimeshaded/PackageListGenerator.java [hunk 1]
```diff
@@ -66,7 +66,8 @@
         "org/apache/xerces",
         "org/w3c/dom",
         "org/xml/sax",
-        "sun/misc");
+        "sun/misc"
+    );
```
Comment (eskatos) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/runtimeshaded/PackageListGenerator.java]: :o: unrelated change could be removed from this PR

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java [hunk 2]
```diff
@@ -61,17 +60,10 @@ public FileCollectionResolveContext add(Object element) {
...
-            Object closureResult = closure.call();
-            if (closureResult != null) {
-                add(closureResult);
-            }
         } else if (element instanceof Callable) {
-            Callable callable = (Callable) element;
-            Object callableResult = uncheckedCall(callable);
-            if (callableResult != null) {
-                add(callableResult);
+            Object deferredResult = uncheckedCall((Callable) element);
```
Comment (eskatos) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java]: ❓  shouldn't this use `DeferredUtil`?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java [hunk 3]
```diff
@@ -55,17 +54,10 @@ public void convert(Object notation, NotationConvertResult<? super String> resul
                 || notation instanceof Number
                 || notation instanceof Boolean) {
             result.converted(notation.toString());
-        } else if (notation instanceof Closure) {
-            final Closure closure = (Closure) notation;
-            final Object called = closure.call();
-            convert(called, result);
         } else if (notation instanceof Callable) {
             final Callable callableNotation = (Callable) notation;
             final Object called = uncheckedCall(callableNotation);
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java]: ❓ shouldn't this use `DeferredUtil`?

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java [hunk 4]
```diff
@@ -121,16 +119,15 @@ public final ResolvableFileCollectionResolveContext newContext() {
             } else if (element instanceof TaskOutputs) {
                 TaskOutputs outputs = (TaskOutputs) element;
                 queue.add(0, outputs.getFiles());
-            } else if (element instanceof Callable) {
-                Callable callable = (Callable) element;
-                Object callableResult = uncheckedCall(callable);
-                if (callableResult != null) {
-                    queue.add(0, callableResult);
-                }
             } else if (element instanceof Provider) {
```
Comment (eskatos) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java]: :x: this branch is handled by `DeferredUtil` below and should be removed

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java (hunk 2 above)
Comment (bamboo) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java]: That would change the behaviour for `Provider<T>` values.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java (hunk 3 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java]: That would change the behaviour for `Provider<T>` values.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/runtimeshaded/PackageListGenerator.java (hunk 1 above)
Comment (bamboo) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/runtimeshaded/PackageListGenerator.java]: Thanks.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java (hunk 4 above)
Comment (bamboo) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java]: Right. The comparison to `null` made me postpone this but upon closer inspection it makes sense. Thanks.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java (hunk 2 above)
Comment (eskatos) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java]: Right, and providers here should be handled separately to visit their build dependencies. Out of this PR's scope.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java (hunk 3 above)
Comment (eskatos) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java]: Right. It would add support for `Provider<T>` where `T` is supported. I think we want it on `PathNotationConverter` for uniformity and to support e.g.:
```kotlin
copy {
    from { "foo/dir" }
    into { "bar/dir" }
}
```

Can be considered out of this PR's scope and done as a subsequent step for https://github.com/gradle/kotlin-dsl/issues/1077.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java (hunk 2 above)
Comment (bamboo) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java]: Right.

**Comment by bamboo** [comment]
> Shouldn't DefaultTaskDependency#visitDependencies also use DeferredUtil?

`visitDependencies` is more complicated as it handles `Closure` specially:

https://github.com/gradle/gradle/blob/c23e6a2b40ab45c3e4e1ca22bb0ad1078a30943d/subprojects/model-core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskDependency.java#L103-L104

And `Provider` is not handled explicitly at all.

Better left for the future if we decide to accept Kotlin lambdas as task dependencies.
